### ISSUE TITLE
Interesting exception trying to create shortcuts

### ISSUE THREAD

**Issue by chrisalexander** [Squirrel/Squirrel.Windows#1336]
I noticed our app updated and all our shortcuts disappeared on numerous machines. I have Squirrel detailed logging turned on. When I re-installed the app, I get the below error and log message:

Log:
```
Can't write shortcut: C:\Users\my.user\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\company\appname.lnk
```

Exception:

```
mscorlib : System.Reflection.TargetParameterCountException
Parameter count mismatch.
   at System.Reflection.RuntimeMethodInfo.InvokeArgumentsCheck(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Splat.NLog.NLogSplatLogger.Info(String message, Object[] args)
   at Squirrel.UpdateManager.ApplyReleasesImpl.<>c__DisplayClass5_0.<CreateShortcutsForExecutable>b__1() in C:\Users\paulb\code\Squirrel\squirrel.windows\src\Squirrel\UpdateManager.ApplyReleases.cs:line 249
   at Squirrel.Utility.<>c__DisplayClass8_0.<Retry>b__0() in C:\Users\paulb\code\Squirrel\squirrel.windows\src\Squirrel\Utility.cs:line 142
   at Squirrel.Utility.Retry[T](Func`1 block, Int32 retries) in C:\Users\paulb\code\Squirrel\squirrel.windows\src\Squirrel\Utility.cs:line 161
   at Squirrel.Utility.LogIfThrows(IFullLogger This, LogLevel level, String message, Action block) in C:\Users\paulb\code\Squirrel\squirrel.windows\src\Squirrel\Utility.cs:line 575
```

The preceeding log message is:
```
Creating shortcut for app.exe => C:\Users\my.user\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\company\appname.lnk
```

This leads me to believe it's dying somewhere in [here](https://github.com/Squirrel/Squirrel.Windows/blob/b25c9fa1e4d60f8969137532a95bd53bb5a8de1f/src/Squirrel/UpdateManager.ApplyReleases.cs) between line 227 and 246 (inclusive) though exactly which call is failing I can't determine by static analysis. I think this fails first and thus explains why desktop shortcuts aren't created, the next one to try in the for-loop due to the ordering of the enum [here](https://github.com/Squirrel/Squirrel.Windows/blob/8877944cf21a58c18341d6640bed2014582d1380/src/Squirrel/IUpdateManager.cs#L16). Future updates don't create the shortcuts due to the check on line 218 of the above file.

Interestingly, when an update is applied once the shortcuts are deleted, the line 218 check means the start menu shortcut is skipped, and then the desktop one is tried to be created, but results in the same exception.

This is as far as I have got with the sleuthing, I apologise I haven't been able to work out exactly which line is causing the problem. I am using .NET 4.6.1 and Squirrel 1.8,0.

Cheers

Chris

**Comment by chrisalexander** [comment]
I have traced this down to the log in UpdateManager.ApplyReleases.cs line 248, which led me to check which nuget packages I have installed. I have Splat 1.6.2 (which is hugely out of date but I think what is required for Squirrel still) and Splat.NLog 1.0.5 (so I can redirect the Squirrel log messages to NLog). These haven't changed for quite a while. Then I noticed I had upgraded NLog from 4.5.2 to 4.5.6, so I tried rolling that back, which didn't sort. Then I rolled Squirrel back to 1.7.9, which was the last version that worked (in addition to all previous), and that also didn't work. This implies I hadn't noticed it not working before. I uninstalled Splat.NLog and that sorted it out.

I then re-installed Splat.NLog and move the `Splat.NLog.Helpers.UseNLog();` code to after the Squirrel initialisation, and it continued to work. So I undid all the above and just made this change (moving the splat nlog helper) and it appears to now work as expected.

Unfortunately, I had added this code so that I could get diagnostic information about squirrel updates, which were failing and I needed to know how it had got on!

So I think the issue can be summarised to, "Squirrel can't create shortcuts when Splat logging is redirected to NLog". I am leaving this open until a maintainer has a chance to look at it and decide what best to do about it. It might be classed as a Splat issue?
